### PR TITLE
fix samples and add a note to doxygen of c_type fn

### PR DESCRIPTION
**Pull request by r-rojo**
Fix the C++ sample for hello_fpga-1.cpp to use explicit types instead of
the `auto` keyword. This is done to make it clearer as to the types of
objects being created.

Add a note in shared_buffer::c_type doxygen markup to warn developers of
the dangers of _stealing_ a raw pointer from the shared_buffer object.

### CODE REVIEW

File: common/include/opae/cxx/core/shared_buffer.h [hunk 1]
```diff
@@ -87,6 +87,10 @@ class shared_buffer {
   void release();
 
   /** Retrieve the virtual address of the buffer base.
+   *
+   *  @note assiging this to a variable should be done in limited scopes
```
Comment (tswhison) [inline common/include/opae/cxx/core/shared_buffer.h]: s/assiging/assigning/

File: common/include/opae/cxx/core/shared_buffer.h [hunk 2]
```diff
@@ -87,6 +87,10 @@ class shared_buffer {
...
+   *  as the variable can outlive the shared_buffer object creating the danger
+   *  of a dangling pointer
```
Comment (tswhison) [inline common/include/opae/cxx/core/shared_buffer.h]: period at end?

**Comment by michael-adler** [comment]
Your note will be interpreted well by C++ experts, but I'm less convinced that casual users will understand the implications. How about explaining that the the buffer pointer is a smart pointer, that the destructor is called automatically when the last reference is deleted (e.g. when leaving the current scope) and that the pointers that result from reinterpret_cast are not smart and not part of the reference count?

File: common/include/opae/cxx/core/shared_buffer.h [hunk 3]
```diff
@@ -88,9 +88,17 @@ class shared_buffer {
 
   /** Retrieve the virtual address of the buffer base.
    *
-   *  @note assiging this to a variable should be done in limited scopes
-   *  as the variable can outlive the shared_buffer object creating the danger
-   *  of a dangling pointer
+   *  @note Instances of a shared buffer can only be created using either
```
Comment (tswhison) [inline common/include/opae/cxx/core/shared_buffer.h]: maybe an example to clarify/drive home the point?

Comment (r-rojo) [reply common/include/opae/cxx/core/shared_buffer.h]: Example here? The cxx hello fpga isn't enough?

Comment (tswhison) [reply common/include/opae/cxx/core/shared_buffer.h]: Maybe offer up the issue submitter's code as an example of de-referencing the invalid pointer after the scope of the shared_ptr..  Having seen the code, it was clear to me what you were describing.  I'm concerned that others might not get the point by only reading the prose.

**Pull request merged**